### PR TITLE
EHR Sample Manager Module

### PR DESCRIPTION
**Pull request by labkey-martyp**
#### Rationale
This can be enabled within a Sample Manager project on a server with an EHR folder (module properties defined at site level), to foreign key in data from the EHR for an Animal Source.  This data will be foreign keyed to Animal source and samples with a single animal source (across full lineage).  Allows filtering/sorting on foreign keyed data from EHR container.

#### Related Pull Requests
* https://github.com/LabKey/platform/pull/3080
* https://github.com/LabKey/sampleManagement/pull/850

#### Changes
* Create EHR_SM module
* Animal ETL and .qview.xmls to populate Animal source and define some views
* Java customizer to setup Animal foreign key to the EHR folder using EHR site level module properties
* Java customizer to setup samples tables with foreign key to animal record associated with animal source of a sample
* Age at sample time column to calculate animal age when sample was taken
* Admin page to select which sample types are from animal sources and which sample date column to use for animal age calculation

### CODE REVIEW

File: EHR_SM/src/org/labkey/ehr_sm/EHR_SMContainerListener.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import org.jetbrains.annotations.NotNull;
+import org.labkey.api.data.Container;
+import org.labkey.api.data.ContainerManager.ContainerListener;
+import org.labkey.api.security.User;
+import java.util.Collections;
+import java.util.Collection;
+
+import java.beans.PropertyChangeEvent;
+
+public class EHR_SMContainerListener implements ContainerListener
```
Comment (labkey-jeckels) [inline EHR_SM/src/org/labkey/ehr_sm/EHR_SMContainerListener.java:28]: Safe to delete this?

File: EHR_SM/src/org/labkey/ehr_sm/EHR_SMCustomizer.java [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+                    public DisplayColumn createRenderer(ColumnInfo colInfo)
+                    {
+                        return new DataColumn(colInfo)
+                        {
+
+                            @Override
+                            public Object getValue(RenderContext ctx)
+                            {
+                                String value = (String) super.getValue(ctx);
+                                if (value != null && value.startsWith("#Error"))
```
Comment (labkey-jeckels) [inline EHR_SM/src/org/labkey/ehr_sm/EHR_SMCustomizer.java]: If we stick with handling multiple values with this string constant, can we use a constant somewhere?

File: EHR_SM/build.gradle [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import org.labkey.gradle.util.BuildUtils;
+
+plugins {
+    id 'org.labkey.build.module'
+}
+
+dependencies {
+    BuildUtils.addLabKeyDependency(project: project, config: "implementation", depProjectPath: ":server:modules:dataintegration", depProjectConfig: "apiJarFile")
```
Comment (labkey-jeckels) [inline EHR_SM/build.gradle:8]: Do we need this dependency on the API JAR?

**Review by labkey-jeckels — APPROVED** [review]

**Comment by labkey-martyp** [comment]
https://github.com/LabKey/ehrModules/pull/422

**Pull request closed without merging**